### PR TITLE
Document most of the Topiary library's public and private symbols

### LINKED ISSUE #320: Ensure library source files are adequately documented
> As we are targeting a library use-case at some point, it would make sense to start writing documentation in the module headers. For now, this is good though!

_Originally posted by @ErinvanderVeen in https://github.com/tweag/topiary/pull/315#discussion_r1117230804_

### PR DESCRIPTION
**Pull request by ErinvanderVeen**
Resolves #320

### CODE REVIEW

File: topiary/src/atom_collection.rs [hunk 1]
```diff
@@ -9,26 +9,57 @@ use tree_sitter_facade::Node;
...
+    /// something to a node, a new Atom is added to this HashMap.
+    /// The key of the hashmap is the identifier of the node.
     append: HashMap<usize, Vec<Atom>>,
+    /// A query file can define custom leaf nodes (nodes that Topiary should not
+    /// touch during formatting). When such a node is encountered, its id is stored in
+    /// this HashSet.
     specified_leaf_nodes: HashSet<usize>,
+    /// If a node is a leaf, or if it is explicitly marked as such by the
+    /// formatting directives, it is added to this HashMap as the key. The value
+    /// of the Map contains all child nodes.
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:44]:
```suggestion
    /// of the Map contains all parent nodes.
```

File: topiary/src/atom_collection.rs [hunk 2]
```diff
@@ -9,26 +9,57 @@ use tree_sitter_facade::Node;
...
     parent_leaf_nodes: HashMap<usize, usize>,
+    /// Topiary has some formatting directives that only apply if a node spans
+    /// multiple lines. During initial collection all such nodes are added to this
+    /// HashSet for easy checking if a node spans multiple lines.
     multi_line_nodes: HashSet<usize>,
+    /// During initial Atom collection, any node that has a blank lines above
+    /// the node is added to this HashSet.
     blank_lines_before: HashSet<usize>,
+    /// During initial Atom collection, any node that has a linebreak directly
+    /// before it the node is added to this HashSet.
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:54]:
```suggestion
    /// before it is added to this HashSet.
```

File: topiary/src/atom_collection.rs [hunk 3]
```diff
@@ -9,26 +9,57 @@ use tree_sitter_facade::Node;
...
     line_break_before: HashSet<usize>,
+    /// During initial Atom collection, any node that has a linebreak directly
+    /// after it the node is added to this HashSet.
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:57]:
```suggestion
    /// after it is added to this HashSet.
```

File: topiary/src/atom_collection.rs [hunk 4]
```diff
@@ -93,7 +124,21 @@ impl AtomCollection {
         }
     }
 
-    /// This gets called a lot during query processing, and needs to be efficient.
+    /// Resolves a capture name by modifying the AtomCollection based on the
+    /// instructions provided by the capture name on the Node.
+    ///
+    ///
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:130]: Double blank line.

File: topiary/src/atom_collection.rs [hunk 5]
```diff
@@ -93,7 +124,21 @@ impl AtomCollection {
...
+    /// * `name` - The name of the capture, starting with `@`.
+    /// * `node` - The node that matches the capture in the syntax tree.
+    /// * `predicates` - The query predicates that modify the formatting behavior for the capture.
+    ///
+    /// # Errors
+    ///
+    /// This function returns an error if the capture name requires a predicate that is not present.
+    /// It also returns an error if the capture name is not recognized by Topiary.
+    // NOTE: This gets called a lot during query processing, and needs to be efficient.
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:141]: Perhaps upgrade this comment?

File: topiary/src/atom_collection.rs [hunk 6]
```diff
@@ -308,13 +353,30 @@ impl AtomCollection {
...
         self.parent_leaf_nodes.insert(node.id(), parent_id);
         for child in node.children(&mut node.walk()) {
             self.mark_leaf_parent(&child, parent_id);
         }
     }
 
+    /// This function collects the leaf nodes of a tree-sitter CST tree and stores them in the `atoms` field of the internal formatter state.
+    /// It also marks the leaf parent of each node in the `leaf_parent` field of the formatter.
+    /// A leaf node is either a node with no children or a node that is specified as a leaf node by the formatter.
+    /// A leaf parent is the closest ancestor of a node that is a leaf node.
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:368]:
```suggestion
    /// A leaf parent is the closest ancestor of a leaf node.
```

File: topiary/src/atom_collection.rs [hunk 7]
```diff
@@ -404,6 +492,23 @@ impl AtomCollection {
             });
     }
 
+    /// Expands a softline atom to a hardline, space or empty atom depending on
+    /// if we are in a multiline context or not.
+    ///
+    /// If the node's parent is labelled as a multi-line node, it expands the
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:498]: Sometimes you have written `node`'s

File: topiary/src/atom_collection.rs [hunk 8]
```diff
@@ -720,16 +857,34 @@ impl AtomCollection {
...
+    /// `@append_delimiter` and `@prepend_delimiter` directives.
     pub delimiter: Option<String>,
+    /// The predicate used to name a scope indicated by `@begin_scope` and `@end_scope`.
     pub scope_id: Option<String>,
+    /// The flag that indicates whether the query only matches single-line nodes.
     pub single_line_only: bool,
+    /// The flag that indicates whether the query only matches multi-line nodes.
     pub multi_line_only: bool,
+    /// The flag that indicates that the query only triggers if the associated
+    /// custom scope containing the matched nodes are is single-line.
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:872]:
```suggestion
    /// custom scope containing the matched nodes is single-line.
```

File: topiary/src/atom_collection.rs [hunk 9]
```diff
@@ -720,16 +857,34 @@ impl AtomCollection {
...
     pub single_line_scope_only: Option<String>,
+    /// The flag that indicates that the query only triggers if the associated
+    /// custom scope containing the matched nodes are is multi-line.
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:875]:
```suggestion
    /// custom scope containing the matched nodes is multi-line.
```

File: topiary/src/atom_collection.rs [hunk 10]
```diff
@@ -781,6 +949,20 @@ fn dfs_flatten<'tree>(node: &Node<'tree>) -> Vec<Node<'tree>> {
     dfs_nodes
 }
 
+/// Detects multi-line nodes in a vector of nodes and returns a set of their ids.
+///
+/// This function takes a slice of `Node`s that represent the nodes in a depth-first search
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:954]: Sometimes you write `node`'s

Comment (torhovland) [inline topiary/src/atom_collection.rs:954]:
```suggestion
/// This function takes a slice of `Node`s that represents the nodes in a depth-first search
```

File: topiary/src/configuration.rs [hunk 11]
```diff
@@ -1,8 +1,15 @@
+/// Topiary can be configured using the `Configuration` struct.
+/// A basic configuration, written in toml, it included buildtime and parsed runtime.
```
Comment (torhovland) [inline topiary/src/configuration.rs:2]:
```suggestion
/// A basic configuration, written in toml, it is included buildtime and parsed runtime.
```

File: topiary/src/configuration.rs [hunk 12]
```diff
@@ -1,8 +1,15 @@
...
+/// Additional configuration has to be provided by the user of the library.
 use std::{collections::HashSet, str::from_utf8};
 
 use crate::{language::Language, FormatterError, FormatterResult};
 use serde::{Deserialize, Serialize};
 
+/// The configuration of Topiary. Contains information on how to format every language.
+/// Can be provided by the user of the library, alternatively, Topiary ships with a default
```
Comment (torhovland) [inline topiary/src/configuration.rs:10]:
```suggestion
/// Can be provided by the user of the library, or alternatively, Topiary ships with a default
```

File: topiary/src/configuration.rs [hunk 13]
```diff
@@ -13,6 +20,8 @@ impl Configuration {
         Configuration { language: vec![] }
     }
 
+    /// Collects the known extensions of all languages into a single HashSet.
+    /// Usefull for testing if Topiary is able to configure the given file.
```
Comment (torhovland) [inline topiary/src/configuration.rs:24]:
```suggestion
    /// Useful for testing if Topiary is able to configure the given file.
```

File: topiary/src/configuration.rs [hunk 14]
```diff
@@ -24,6 +33,11 @@ impl Configuration {
         res
     }
 
+    /// Gets a language configuration from the entire configuration.
+    ///
+    /// # Errors
+    /// If the provided language name cannot be found in the Configuration, this
```
Comment (torhovland) [inline topiary/src/configuration.rs:39]: Needs a blank line after heading

File: topiary/src/configuration.rs [hunk 15]
```diff
@@ -35,6 +49,9 @@ impl Configuration {
         ));
     }
 
+    /// Parse the default configuration directly into a `Configuration`,
+    /// This is usefull for users of Topiary that have no special requirements.
```
Comment (torhovland) [inline topiary/src/configuration.rs:53]:
```suggestion
    /// This is useful for users of Topiary that have no special requirements.
```

File: topiary/src/configuration.rs [hunk 16]
```diff
@@ -35,6 +49,9 @@ impl Configuration {
...
+    /// It is also incredibly usefull in tests.
```
Comment (torhovland) [inline topiary/src/configuration.rs:54]:
```suggestion
    /// It is also incredibly useful in tests.
```

File: topiary/src/lib.rs [hunk 17]
```diff
@@ -97,9 +97,13 @@ pub enum Atom {
     },
 }
 
+/// Used in `Atom::ScopedConditional` to apply the containing Atoms only if
+/// the matched node spans a single line or multiple lines
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
 pub enum ScopeCondition {
+    /// The Atom is only applied the the matching node spans exactly one line
```
Comment (torhovland) [inline topiary/src/lib.rs:104]:
```suggestion
    /// The Atom is only applied if the matching node spans exactly one line
```

File: topiary/src/lib.rs [hunk 18]
```diff
@@ -97,9 +97,13 @@ pub enum Atom {
...
     SingleLineOnly,
+    /// The Atom is only applied the the matching node spans two or more lines
```
Comment (torhovland) [inline topiary/src/lib.rs:106]:
```suggestion
    /// The Atom is only applied if the matching node spans two or more lines
```

File: topiary/src/pretty.rs [hunk 19]
```diff
@@ -1,7 +1,19 @@
+//! After being split into Atoms, and the querries having been applied this
```
Comment (torhovland) [inline topiary/src/pretty.rs:1]:
```suggestion
//! After being split into Atoms, and the queries having been applied this
```

**Review by torhovland — APPROVED** [review]
Great work!

File: topiary/src/atom_collection.rs (hunk 7 above)
Comment (ErinvanderVeen) [reply topiary/src/atom_collection.rs:502]: I've renamed all `node`'s to `Node`'s.

**Pull request merged**